pyogrio/tests/test_geopandas_io.py

```diff
@@ -259,6 +259,44 @@ def test_write_empty_dataframe(tmpdir, driver, ext):
     assert_geodataframe_equal(df, expected)
 
 
+@pytest.mark.parametrize(
+    "driver,ext", [("GeoJSON", "geojson"), ("GPKG", "gpkg"), ("FlatGeobuf", "fbg")]
+)
+def test_write_mixed_geometries(tmp_path, driver, ext):
+    from shapely.geometry import Point, LineString, box
+
+    df = gp.GeoDataFrame(
+        {"col": [1.0, 2.0, 3.0]},
+        geometry=[Point(0, 0), LineString([(0, 0), (1, 1)]), box(0, 0, 1, 1)],
+        crs="EPSG:4326"
+    )
+
+    filename = tmp_path / f"test.{ext}"
+    write_dataframe(df, filename, driver=driver)
+
+    assert read_info(filename)["geometry_type"] == "Unknown"
+    result = read_dataframe(filename)
+    if driver == "FlatGeobuf":
+        # FlatGeobuf results in mixed row order in case of mixed geometries
+        result = result.sort_values("col").reset_index(drop=True)
+    assert_geodataframe_equal(result, df)
+
+
+def test_write_mixed_geometries_unsupported(tmp_path):
+    # Shapefile doesn't support generic "Geometry" / "Unknown" type
+    # for mixed geometries
+    from shapely.geometry import Point, LineString, box
+
+    df = gp.GeoDataFrame(
+        {"col": [1.0, 2.0, 3.0]},
+        geometry=[Point(0, 0), LineString([(0, 0), (1, 1)]), box(0, 0, 1, 1)],
+        crs="EPSG:4326"
+    )
+
+    with pytest.raises(FeatureError):
```

Comment:
```suggestion
    with pytest.raises(FeatureError, match="Could not add feature to layer"):
```

Reply:
I added this suggestion, but I also added a TODO that we should propagate better the (more informative) error message from GDAL ("Attempt to write non-point (LINESTRING) geometry to point shapefile")